File: ProblemSolve/71.cpp
/*****************
 * Simplify Path *
 *****************/

class Solution {
public:
    vector<string> split(string s, char delimiter){
        vector<string> splited;
        stringstream ss(s);
        string temp;
    
        while (getline(ss, temp, delimiter)) {
            splited.push_back(temp);
        }
 
        return splited;
    }
    
    string simplifyPath(string path) {
        stack<string> pathStack;
        vector<string> paths = split(path, '/');
        
        for(int i=0; i<paths.size(); ++i){
            if(paths[i] == ".."){
                if(!pathStack.empty()){
                    pathStack.pop();
                }
                continue;
            }
            
            if(paths[i] == "." || paths[i] == ""){
                continue;
            }
            
            pathStack.push(paths[i]);
        }
        
        string ans = "";
        
        while(!pathStack.empty()){
            string p = pathStack.top();
            pathStack.pop();
            if(p == "/"){
                continue;
            }
            
            ans = "/" + p + ans;
        }
        
        if(ans.length() != 0 && ans[ans.length() - 1] == '/'){
            ans.pop_back();
        }
        
        if(ans.length() == 0){
            return "/";
        }
        
        return ans;
    }
};

File: ProblemSolve/71.cs
/*****************
 * Simplify Path *
 *****************/

public class Solution {
    public string SimplifyPath(string path) {
        string[] paths = path.Split('/');
        Stack<string> pathBuilder = new Stack<string>();
        
        for(int i=0; i<paths.Length; i++){
            if(paths[i].Equals("..")){
                if(pathBuilder.Count != 0){
                    pathBuilder.Pop();
                }
                
                continue;
            }
            
            if(paths[i].Equals(".") || paths[i].Equals("")){
                continue;
            }

            pathBuilder.Push(paths[i]);
        }
        
        StringBuilder ans = new StringBuilder();
        
        while(pathBuilder.Count != 0){
            string p = pathBuilder.Pop();
            if(p.Equals("/")){
                continue;
            }
            
            ans.Insert(0, "/" + p);
        }
        
        if(ans.Length != 0 && ans[ans.Length - 1] == '/'){
            ans.Remove(ans.Length-1, 1);
        }
        
        if(ans.Length == 0){
            return "/";
        }
        
        return ans.ToString();
    }
}